ethcore/src/state/mod.rs

```diff
@@ -539,8 +542,55 @@ impl<B: Backend> State<B> {
 			|a| a.as_ref().and_then(|account| account.storage_root().cloned()))
 	}
 
-	/// Mutate storage of account `address` so that it is `value` for `key`.
-	pub fn storage_at(&self, address: &Address, key: &H256) -> TrieResult<H256> {
+	/// Get the value of storage at a specific checkpoint.
+	pub fn checkpoint_storage_at(&self, checkpoint_index: usize, address: &Address, key: &H256) -> TrieResult<Option<H256>> {
```

Comment:
I'm unconvinced this code actually works correctly, do we have some higher-level tests to prove me wrong?

Best with scenarios, where we have deeply nested calls, the accounts are suicided in the meantime, the values are changed multiple times on different levels, etc.